lib/libc/sys/sync.2
.\"	$NetBSD: sync.2,v 1.17.36.1 2017/05/19 00:22:52 pgoyette Exp $
.\"
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)sync.2	8.1 (Berkeley) 6/4/93
.\"
.Dd March 25, 2009
.Dt SYNC 2
.Os
.Sh NAME
.Nm sync
.Nd "synchronize disk block in-core status with that on disk"
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In unistd.h
.Ft void
.Fn sync void
.Sh DESCRIPTION
The
.Fn sync
function forces a write of dirty (modified) buffers
in the block buffer cache out
to disk.
The kernel keeps this information in core to reduce
the number of disk I/O transfers required by the system.
As information in the cache is lost after a system crash,
kernel thread
.Nm ioflush
ensures that dirty buffers are synced to disk
eventually.
By default, a dirty buffer is synced after 30 seconds,
but some file systems exploit
.Nm ioflush
features to sync directory data and metadata faster
(after 15 and 10 seconds, respectively).
.Pp
The function
.Xr fsync 2
may be used to synchronize individual file descriptor
attributes.
.Sh CAUTIONS
Many modern disks contain write-back caches.
In theory
.Fn sync
flushes these.
In practice there are many possible ways for this mechanism to go
astray.
It is prudent (where possible) to allow a few seconds after syncing
for everything to settle before e.g. turning off the power.
.Pp
It may also be desirable to use
.Xr dkctl 8
or
.Xr scsictl 8
to disable the write-back cache entirely.
.Sh SEE ALSO
.Xr fsync 2 ,
.Xr dkctl 8 ,
.Xr scsictl 8 ,
.Xr sync 8
.Sh HISTORY
A
.Fn sync
function call appeared in
.At v6 .
.Pp
Historically,
.Fn sync
would schedule buffers for writing but not actually wait for the
writes to finish.
It was necessary to issue a second or sometimes a third call to ensure
that all buffers had in fact been written out.
In
.Nx ,
.Fn sync
does not return until all buffers have been written.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2285 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Change arc4random(3) to use ChaCha20 and per-thread state.
		[riastradh 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]

games/sail/sync.c
/*	$NetBSD: sync.c,v 1.33.12.1 2014/08/20 00:00:23 tls Exp $	*/

/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
#if 0
static char sccsid[] = "@(#)sync.c	8.2 (Berkeley) 4/28/95";
#else
__RCSID("$NetBSD: sync.c,v 1.33.12.1 2014/08/20 00:00:23 tls Exp $");
#endif
#endif /* not lint */

#include <sys/stat.h>

#include <fcntl.h>
#include <errno.h>
#include <limits.h>
#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include "extern.h"
#include "pathnames.h"

#define BUFSIZE 4096

/* Message types */
#define W_CAPTAIN	1
#define W_CAPTURED	2
#define W_CLASS		3
#define W_CREW		4
#define W_DBP		5
#define W_DRIFT		6
#define W_EXPLODE	7
/*      W_FILE		8   not used */
#define W_FOUL		9
#define W_GUNL		10
#define W_GUNR		11
#define W_HULL		12
#define W_MOVE		13
#define W_OBP		14
#define W_PCREW		15
#define W_UNFOUL	16
#define W_POINTS	17
#define W_QUAL		18
#define W_UNGRAP	19
#define W_RIGG		20
#define W_COL		21
#define W_DIR		22
#define W_ROW		23
#define W_SIGNAL	24
#define W_SINK		25
#define W_STRUCK	26
#define W_TA		27
#define W_ALIVE		28
#define W_TURN		29
#define W_WIND		30
#define W_FS		31
#define W_GRAP		32
#define W_RIG1		33
#define W_RIG2		34
#define W_RIG3		35
#define W_RIG4		36
#define W_BEGIN		37
#define W_END		38
#define W_DDEAD		39


static void recv_captain(struct ship *ship, const char *astr);
static void recv_captured(struct ship *ship, long a);
static void recv_class(struct ship *ship, long a);
static void recv_crew(struct ship *ship, long a, long b, long c);
static void recv_dbp(struct ship *ship, long a, long b, long c, long d);
static void recv_drift(struct ship *ship, long a);
static void recv_explode(struct ship *ship, long a);
static void recv_foul(struct ship *ship, long a);
static void recv_gunl(struct ship *ship, long a, long b);
static void recv_gunr(struct ship *ship, long a, long b);
static void recv_hull(struct ship *ship, long a);
static void recv_move(struct ship *ship, const char *astr);
static void recv_obp(struct ship *ship, long a, long b, long c, long d);
static void recv_pcrew(struct ship *ship, long a);
static void recv_unfoul(struct ship *ship, long a, long b);
static void recv_points(struct ship *ship, long a);
static void recv_qual(struct ship *ship, long a);
static void recv_ungrap(struct ship *ship, long a, long b);
static void recv_rigg(struct ship *ship, long a, long b, long c, long d);
static void recv_col(struct ship *ship, long a);
static void recv_dir(struct ship *ship, long a);
static void recv_row(struct ship *ship, long a);
static void recv_signal(struct ship *ship, const char *astr);
static void recv_sink(struct ship *ship, long a);
static void recv_struck(struct ship *ship, long a);
static void recv_ta(struct ship *ship, long a);
static void recv_alive(void);
static void recv_turn(long a);
static void recv_wind(long a, long b);
static void recv_fs(struct ship *ship, long a);
static void recv_grap(struct ship *ship, long a);
static void recv_rig1(struct ship *ship, long a);
static void recv_rig2(struct ship *ship, long a);
static void recv_rig3(struct ship *ship, long a);
static void recv_rig4(struct ship *ship, long a);
static void recv_begin(struct ship *ship);
static void recv_end(struct ship *ship);
static void recv_ddead(void);

static void Write(int, struct ship *, long, long, long, long);
static void Writestr(int, struct ship *, const char *);

static int sync_update(int, struct ship *, const char *,
		       long, long, long, long);

static char sync_buf[BUFSIZE];
static char *sync_bp = sync_buf;
static long sync_seek;
static FILE *sync_fp;

static const char *
get_sync_file(int scenario_number)
{
	static char sync_file[NAME_MAX];

	snprintf(sync_file, sizeof(sync_file), _FILE_SYNC, scenario_number);
	return sync_file;
}

static const char *
get_lock_file(int scenario_number)
{
	static char sync_lock[NAME_MAX];

	snprintf(sync_lock, sizeof(sync_lock), _FILE_LOCK, scenario_number);
	return sync_lock;
}

void
fmtship(char *buf, size_t len, const char *fmt, struct ship *ship)
{
	while (*fmt) {
		if (len-- == 0) {
			*buf = '\0';
			return;
		}
		if (*fmt == '$' && fmt[1] == '$') {
			size_t l = snprintf(buf, len, "%s (%c%c)",
			    ship->shipname, colours(ship), sterncolour(ship));
			buf += l;
			len -= l - 1;
			fmt += 2;
		}
		else
			*buf++ = *fmt++;
	}

	if (len > 0)
		*buf = '\0';
}


/*VARARGS3*/
void
makesignal(struct ship *from, const char *fmt, struct ship *ship, ...)
{
	char message[BUFSIZ];
	char format[BUFSIZ];
	va_list ap;

	va_start(ap, ship);
	fmtship(format, sizeof(format), fmt, ship);
	vsnprintf(message, sizeof(message), format, ap);
	va_end(ap);
	send_signal(from, message);
}

/*VARARGS2*/
void
makemsg(struct ship *from, const char *fmt, ...)
{
	char message[BUFSIZ];
	va_list ap;

	va_start(ap, fmt);
	vsnprintf(message, sizeof(message), fmt, ap);
	va_end(ap);
	send_signal(from, message);
}

int
sync_exists(int gamenum)
{
	const char *path;
	struct stat s;
	time_t t;

	path = get_sync_file(gamenum);
	time(&t);
	setegid(egid);
	if (stat(path, &s) < 0) {
		setegid(gid);
		return 0;
	}
	if (s.st_mtime < t - 60*60*2) {		/* 2 hours */
		unlink(path);
		path = get_lock_file(gamenum);
		unlink(path);
		setegid(gid);
		return 0;
	} else {
		setegid(gid);
		return 1;
	}
}

int
sync_open(void)
{
	const char *sync_file;
	struct stat tmp;

	if (sync_fp != NULL)
		fclose(sync_fp);
	sync_file = get_sync_file(game);
	(void)get_lock_file(game);
	setegid(egid);
	if (stat(sync_file, &tmp) < 0) {
		mode_t omask = umask(002);
		sync_fp = fopen(sync_file, "w+");
		umask(omask);
	} else
		sync_fp = fopen(sync_file, "r+");
	setegid(gid);
	if (sync_fp == NULL)
		return -1;
	sync_seek = 0;
	return 0;
}

void
sync_close(int doremove)
{
	const char *sync_file;

	if (sync_fp != 0)
		fclose(sync_fp);
	if (doremove) {
		sync_file = get_sync_file(game);
		setegid(egid);
		unlink(sync_file);
		setegid(gid);
	}
}

static void
Write(int type, struct ship *ship, long a, long b, long c, long d)
{
	size_t max = sizeof(sync_buf) - (sync_bp - sync_buf);
	int shipindex = (ship == NULL) ? 0 : ship->file->index;

	snprintf(sync_bp, max, "%d %d 0 %ld %ld %ld %ld\n",
		       type, shipindex, a, b, c, d);
	while (*sync_bp++)
		;
	sync_bp--;
	if (sync_bp >= &sync_buf[sizeof sync_buf])
		abort();
	sync_update(type, ship, NULL, a, b, c, d);
}

static void
Writestr(int type, struct ship *ship, const char *a)
{
	size_t max = sizeof(sync_buf) - (sync_bp - sync_buf);
	int shipindex = (ship == NULL) ? 0 : ship->file->index;

	snprintf(sync_bp, max, "%d %d 1 %s\n", type, shipindex, a);
	while (*sync_bp++)
		;
	sync_bp--;
	if (sync_bp >= &sync_buf[sizeof sync_buf])
		abort();
	sync_update(type, ship, a, 0, 0, 0, 0);
}

int
Sync(void)
{
	sig_t sighup, sigint;
	int n;
	int type, shipnum, isstr;
	char *astr;
	long a, b, c, d;
	char buf[80];
	char erred = 0;
#ifndef LOCK_EX
	const char *sync_file;
	const char *sync_lock;
#endif

	sighup = signal(SIGHUP, SIG_IGN);
	sigint = signal(SIGINT, SIG_IGN);
	for (n = TIMEOUT; --n >= 0;) {
#ifdef LOCK_EX
		if (flock(fileno(sync_fp), LOCK_EX|LOCK_NB) >= 0)
			break;
		if (errno != EWOULDBLOCK)
			return -1;
#else
		sync_file = get_sync_file(game);
		sync_lock = get_lock_file(game);
		setegid(egid);
		if (link(sync_file, sync_lock) >= 0) {
			setegid(gid);
			break;
		}
		setegid(gid);
		if (errno != EEXIST)
			return -1;
#endif
		sleep(1);
	}
	if (n <= 0)
		return -1;
	fseek(sync_fp, sync_seek, SEEK_SET);
	for (;;) {
		switch (fscanf(sync_fp, "%d%d%d", &type, &shipnum, &isstr)) {
		case 3:
			break;
		case EOF:
			goto out;
		default:
			goto bad;
		}
		if (shipnum < 0 || shipnum >= cc->vessels)
			goto bad;
		if (isstr != 0 && isstr != 1)
			goto bad;
		if (isstr) {
			int ch;
			char *p;

			for (p = buf;;) {
				ch = getc(sync_fp);
				*p++ = ch;
				switch (ch) {
				case '\n':
					p--;
				case EOF:
					break;
				default:
					if (p >= buf + sizeof buf)
						p--;
					continue;
				}
				break;
			}
			*p = 0;
			for (p = buf; *p == ' '; p++)
				;
			astr = p;
			a = b = c = d = 0;
		} else {
			if (fscanf(sync_fp, "%ld%ld%ld%ld", &a, &b, &c, &d)
			    != 4)
				goto bad;
			astr = NULL;
		}
		if (sync_update(type, SHIP(shipnum), astr, a, b, c, d) < 0)
			goto bad;
	}
bad:
	erred++;
out:
	if (!erred && sync_bp != sync_buf) {
		fseek(sync_fp, 0L, SEEK_END);
		fwrite(sync_buf, sizeof *sync_buf, sync_bp - sync_buf,
			sync_fp);
		fflush(sync_fp);
		sync_bp = sync_buf;
	}
	sync_seek = ftell(sync_fp);
#ifdef LOCK_EX
	flock(fileno(sync_fp), LOCK_UN);
#else
	setegid(egid);
	unlink(sync_lock);
	setegid(gid);
#endif
	signal(SIGHUP, sighup);
	signal(SIGINT, sigint);
	return erred ? -1 : 0;
}

static int
sync_update(int type, struct ship *ship, const char *astr,
	    long a, long b, long c, long d)
{
	switch (type) {
	case W_CAPTAIN:  recv_captain(ship, astr);    break;
	case W_CAPTURED: recv_captured(ship, a);      break;
	case W_CLASS:    recv_class(ship, a);         break;
	case W_CREW:     recv_crew(ship, a, b, c);    break;
	case W_DBP:      recv_dbp(ship, a, b, c, d);  break;
	case W_DRIFT:    recv_drift(ship, a);         break;
	case W_EXPLODE:  recv_explode(ship, a);       break;
	case W_FOUL:     recv_foul(ship, a);          break;
	case W_GUNL:     recv_gunl(ship, a, b);       break;
	case W_GUNR:     recv_gunr(ship, a, b);       break;
	case W_HULL:     recv_hull(ship, a);          break;
	case W_MOVE:     recv_move(ship, astr);       break;
	case W_OBP:      recv_obp(ship, a, b, c, d);  break;
	case W_PCREW:    recv_pcrew(ship, a);         break;
	case W_UNFOUL:   recv_unfoul(ship, a, b);     break;
	case W_POINTS:   recv_points(ship, a);        break;
	case W_QUAL:     recv_qual(ship, a);          break;
	case W_UNGRAP:   recv_ungrap(ship, a, b);     break;
	case W_RIGG:     recv_rigg(ship, a, b, c, d); break;
	case W_COL:      recv_col(ship, a);           break;
	case W_DIR:      recv_dir(ship, a);           break;
	case W_ROW:      recv_row(ship, a);           break;
	case W_SIGNAL:   recv_signal(ship, astr);     break;
	case W_SINK:     recv_sink(ship, a);          break;
	case W_STRUCK:   recv_struck(ship, a);        break;
	case W_TA:       recv_ta(ship, a);            break;
	case W_ALIVE:    recv_alive();                break;
	case W_TURN:     recv_turn(a);                break;
	case W_WIND:     recv_wind(a, b);             break;
	case W_FS:       recv_fs(ship, a);            break;
	case W_GRAP:     recv_grap(ship, a);          break;
	case W_RIG1:     recv_rig1(ship, a);          break;
	case W_RIG2:     recv_rig2(ship, a);          break;
	case W_RIG3:     recv_rig3(ship, a);          break;
	case W_RIG4:     recv_rig4(ship, a);          break;
	case W_BEGIN:    recv_begin(ship);            break;
	case W_END:      recv_end(ship);              break;
	case W_DDEAD:    recv_ddead();                break;
	default:
		fprintf(stderr, "sync_update: unknown type %d\r\n", type);
		return -1;
	}
	return 0;
}

/*
 * Messages to send
 */

void
send_captain(struct ship *ship, const char *astr)
{
	Writestr(W_CAPTAIN, ship, astr);
}

void
send_captured(struct ship *ship, long a)
{
	Write(W_CAPTURED, ship, a, 0, 0, 0);
}

void
send_class(struct ship *ship, long a)
{
	Write(W_CLASS, ship, a, 0, 0, 0);
}

void
send_crew(struct ship *ship, long a, long b, long c)
{
	Write(W_CREW, ship, a, b, c, 0);
}

void
send_dbp(struct ship *ship, long a, long b, long c, long d)
{
	Write(W_DBP, ship, a, b, c, d);
}

void
send_drift(struct ship *ship, long a)
{
	Write(W_DRIFT, ship, a, 0, 0, 0);
}

void
send_explode(struct ship *ship, long a)
{
	Write(W_EXPLODE, ship, a, 0, 0, 0);
}

void
send_foul(struct ship *ship, long a)
{
	Write(W_FOUL, ship, a, 0, 0, 0);
}

void
send_gunl(struct ship *ship, long a, long b)
{
	Write(W_GUNL, ship, a, b, 0, 0);
}

void
send_gunr(struct ship *ship, long a, long b)
{
	Write(W_GUNR, ship, a, b, 0, 0);
}

void
send_hull(struct ship *ship, long a)
{
	Write(W_HULL, ship, a, 0, 0, 0);
}

void
send_move(struct ship *ship, const char *astr)
{
	Writestr(W_MOVE, ship, astr);
}

void
send_obp(struct ship *ship, long a, long b, long c, long d)
{
	Write(W_OBP, ship, a, b, c, d);
}

void
send_pcrew(struct ship *ship, long a)
{
	Write(W_PCREW, ship, a, 0, 0, 0);
}

void
send_unfoul(struct ship *ship, long a, long b)
{
	Write(W_UNFOUL, ship, a, b, 0, 0);
}

void
send_points(struct ship *ship, long a)
{
	Write(W_POINTS, ship, a, 0, 0, 0);
}

void
send_qual(struct ship *ship, long a)
{
	Write(W_QUAL, ship, a, 0, 0, 0);
}

void
send_ungrap(struct ship *ship, long a, long b)
{
	Write(W_UNGRAP, ship, a, b, 0, 0);
}

void
send_rigg(struct ship *ship, long a, long b, long c, long d)
{
	Write(W_RIGG, ship, a, b, c, d);
}

void
send_col(struct ship *ship, long a)
{
	Write(W_COL, ship, a, 0, 0, 0);
}

void
send_dir(struct ship *ship, long a)
{
	Write(W_DIR, ship, a, 0, 0, 0);
}

void
send_row(struct ship *ship, long a)
{
	Write(W_ROW, ship, a, 0, 0, 0);
}

void
send_signal(struct ship *ship, const char *astr)
{
	Writestr(W_SIGNAL, ship, astr);
}

void
send_sink(struct ship *ship, long a)
{
	Write(W_SINK, ship, a, 0, 0, 0);
}

void
send_struck(struct ship *ship, long a)
{
	Write(W_STRUCK, ship, a, 0, 0, 0);
}

void
send_ta(struct ship *ship, long a)
{
	Write(W_TA, ship, a, 0, 0, 0);
}

void
send_alive(void)
{
	Write(W_ALIVE, NULL, 0, 0, 0, 0);
}

void
send_turn(long a)
{
	Write(W_TURN, NULL, a, 0, 0, 0);
}

void
send_wind(long a, long b)
{
	Write(W_WIND, NULL, a, b, 0, 0);
}

void
send_fs(struct ship *ship, long a)
{
	Write(W_FS, ship, a, 0, 0, 0);
}

void
send_grap(struct ship *ship, long a)
{
	Write(W_GRAP, ship, a, 0, 0, 0);
}

void
send_rig1(struct ship *ship, long a)
{
	Write(W_RIG1, ship, a, 0, 0, 0);
}

void
send_rig2(struct ship *ship, long a)
{
	Write(W_RIG2, ship, a, 0, 0, 0);
}

void
send_rig3(struct ship *ship, long a)
{
	Write(W_RIG3, ship, a, 0, 0, 0);
}

void
send_rig4(struct ship *ship, long a)
{
	Write(W_RIG4, ship, a, 0, 0, 0);
}

void
send_begin(struct ship *ship)
{
	Write(W_BEGIN, ship, 0, 0, 0, 0);
}

void
send_end(struct ship *ship)
{
	Write(W_END, ship, 0, 0, 0, 0);
}

void
send_ddead(void)
{
	Write(W_DDEAD, NULL, 0, 0, 0, 0);
}


/*
 * Actions upon message receipt
 */

static void
recv_captain(struct ship *ship, const char *astr)
{
	strlcpy(ship->file->captain, astr, sizeof ship->file->captain);
}

static void
recv_captured(struct ship *ship, long a)
{
	if (a < 0)
		ship->file->captured = 0;
	else
		ship->file->captured = SHIP(a);
}

static void
recv_class(struct ship *ship, long a)
{
	ship->specs->class = a;
}

static void
recv_crew(struct ship *ship, long a, long b, long c)
{
	struct shipspecs *s = ship->specs;

	s->crew1 = a;
	s->crew2 = b;
	s->crew3 = c;
}

static void
recv_dbp(struct ship *ship, long a, long b, long c, long d)
{
	struct BP *p = &ship->file->DBP[a];

	p->turnsent = b;
	p->toship = SHIP(c);
	p->mensent = d;
}

static void
recv_drift(struct ship *ship, long a)
{
	ship->file->drift = a;
}

static void
recv_explode(struct ship *ship, long a)
{
	if ((ship->file->explode = a) == 2)
		ship->file->dir = 0;
}

static void
recv_foul(struct ship *ship, long a)
{
	struct snag *p = &ship->file->foul[a];

	if (SHIP(a)->file->dir == 0)
		return;
	if (p->sn_count++ == 0)
		p->sn_turn = turn;
	ship->file->nfoul++;
}

static void
recv_gunl(struct ship *ship, long a, long b)
{
	struct shipspecs *s = ship->specs;

	s->gunL = a;
	s->carL = b;
}

static void
recv_gunr(struct ship *ship, long a, long b)
{
	struct shipspecs *s = ship->specs;

	s->gunR = a;
	s->carR = b;
}

static void
recv_hull(struct ship *ship, long a)
{
	ship->specs->hull = a;
}

static void
recv_move(struct ship *ship, const char *astr)
{
	strlcpy(ship->file->movebuf, astr, sizeof ship->file->movebuf);
}

static void
recv_obp(struct ship *ship, long a, long b, long c, long d)
{
	struct BP *p = &ship->file->OBP[a];

	p->turnsent = b;
	p->toship = SHIP(c);
	p->mensent = d;
}

static void
recv_pcrew(struct ship *ship, long a)
{
	ship->file->pcrew = a;
}

static void
recv_unfoul(struct ship *ship, long a, long b)
{
	struct snag *p = &ship->file->foul[a];

	if (p->sn_count > 0) {
		if (b) {
			ship->file->nfoul -= p->sn_count;
			p->sn_count = 0;
		} else {
			ship->file->nfoul--;
			p->sn_count--;
		}
	}
}

static void
recv_points(struct ship *ship, long a)
{
	ship->file->points = a;
}

static void
recv_qual(struct ship *ship, long a)
{
	ship->specs->qual = a;
}

static void
recv_ungrap(struct ship *ship, long a, long b)
{
	struct snag *p = &ship->file->grap[a];

	if (p->sn_count > 0) {
		if (b) {
			ship->file->ngrap -= p->sn_count;
			p->sn_count = 0;
		} else {
			ship->file->ngrap--;
			p->sn_count--;
		}
	}
}

static void
recv_rigg(struct ship *ship, long a, long b, long c, long d)
{
	struct shipspecs *s = ship->specs;

	s->rig1 = a;
	s->rig2 = b;
	s->rig3 = c;
	s->rig4 = d;
}

static void
recv_col(struct ship *ship, long a)
{
	ship->file->col = a;
}

static void
recv_dir(struct ship *ship, long a)
{
	ship->file->dir = a;
}

static void
recv_row(struct ship *ship, long a)
{
	ship->file->row = a;
}

static void
recv_signal(struct ship *ship, const char *astr)
{
	if (mode == MODE_PLAYER) {
		if (nobells)
			Signal("$$: %s", ship, astr);
		else
			Signal("\a$$: %s", ship, astr);
	}
}

static void
recv_sink(struct ship *ship, long a)
{
	if ((ship->file->sink = a) == 2)
		ship->file->dir = 0;
}

static void
recv_struck(struct ship *ship, long a)
{
	ship->file->struck = a;
}

static void
recv_ta(struct ship *ship, long a)
{
	ship->specs->ta = a;
}

static void
recv_alive(void)
{
	alive = 1;
}

static void
recv_turn(long a)
{
	turn = a;
}

static void
recv_wind(long a, long b)
{
	winddir = a;
	windspeed = b;
}

static void
recv_fs(struct ship *ship, long a)
{
	ship->file->FS = a;
}

static void
recv_grap(struct ship *ship, long a)
{
	struct snag *p = &ship->file->grap[a];

	if (SHIP(a)->file->dir == 0)
		return;
	if (p->sn_count++ == 0)
		p->sn_turn = turn;
	ship->file->ngrap++;
}

static void
recv_rig1(struct ship *ship, long a)
{
	ship->specs->rig1 = a;
}

static void
recv_rig2(struct ship *ship, long a)
{
	ship->specs->rig2 = a;
}

static void
recv_rig3(struct ship *ship, long a)
{
	ship->specs->rig3 = a;
}

static void
recv_rig4(struct ship *ship, long a)
{
	ship->specs->rig4 = a;
}

static void
recv_begin(struct ship *ship)
{
	strcpy(ship->file->captain, "begin");
	people++;
}

static void
recv_end(struct ship *ship)
{
	*ship->file->captain = 0;
	ship->file->points = 0;
	people--;
}

static void
recv_ddead(void)
{
	hasdriver = 0;
}
